src/qibo/backends/__init__.py

```diff
@@ -78,39 +79,88 @@ class GlobalBackend(NumpyBackend):
+    def backend(cls):
+        """Get the current backend. If no backend is set, it will create one."""
+        if cls._backend is not None:
+            return cls._backend
+        cls._backend = cls._create_backend()
+        log.info(f"Using {cls._backend} backend on {cls._backend.device}")
+        return cls._backend
 
-        backend = os.environ.get("QIBO_BACKEND")
-        if backend:  # pragma: no cover
+    @classmethod
+    def _create_backend(cls):
+        backend_env = os.environ.get("QIBO_BACKEND")
+        if backend_env:  # pragma: no cover
             # Create backend specified by user
             platform = os.environ.get("QIBO_PLATFORM")
-            cls._instance = construct_backend(backend, platform=platform)
+            backend = construct_backend(backend_env, platform=platform)
         else:
             # Create backend according to default order
             for kwargs in cls._default_order:
                 try:
-                    cls._instance = construct_backend(**kwargs)
+                    backend = construct_backend(**kwargs)
                     break
-                except (ImportError, MissingBackend):
+                except (ModuleNotFoundError, ImportError):
                     pass
 
-        if cls._instance is None:  # pragma: no cover
+        if backend is None:  # pragma: no cover
             raise_error(RuntimeError, "No backends available.")
+        return backend
 
-        log.info(f"Using {cls._instance} backend on {cls._instance.device}")
-        return cls._instance
+    @classmethod
+    def set_backend(cls, backend, **kwargs):
+        cls._backend = construct_backend(backend, **kwargs)
+        log.info(f"Using {cls._backend} backend on {cls._backend.device}")
 
     @classmethod
-    def set_backend(cls, backend, **kwargs):  # pragma: no cover
-        if (
-            cls._instance is None
-            or cls._instance.name != backend
-            or cls._instance.platform != kwargs.get("platform")
-        ):
-            cls._instance = construct_backend(backend, **kwargs)
-        log.info(f"Using {cls._instance} backend on {cls._instance.device}")
+    def transpiler(cls):
+        """Get the current transpiler. If no transpiler is set, it will create one."""
+        if cls._transpiler is not None:
+            return cls._transpiler
+
+        cls._transpiler = cls._default_transpiler()
+        return cls._transpiler
+
+    @classmethod
+    def set_transpiler(cls, transpiler):
+        cls._transpiler = transpiler
+        # TODO: check if transpiler is valid on the backend
+
+    @classmethod
+    def _default_transpiler(cls):  # pragma: no cover
+        import networkx as nx
+
+        from qibo.transpiler.optimizer import Preprocessing
+        from qibo.transpiler.pipeline import Passes
+        from qibo.transpiler.placer import Trivial
+        from qibo.transpiler.router import Sabre
+        from qibo.transpiler.unroller import NativeGates, Unroller
```

Comment:
Yes, moving all of them to the top causes errors.

Reply:
Could you reconstruct the reason?

Reply:
Because of the circular dependency. Would you like to address this now?

Reply:
> Would you like to address this now?

Nope, I'm opening a separate issue for circular dependencies in general.

> Because of the circular dependency.

The part I was interested in was *which* circular dependency (i.e. reconstructing the circle).

Reply:
The loop is roughly structured like this: many modules in `qibo.__init__.py` import the backend at the top of their files. If we import the transpiler at the top of `backend.__init__.py`, it causes a circular import because many transpiler modules rely on `qibo.models.Circuit`, which depends on the backend.

![Untitled Diagram drawio](https://github.com/user-attachments/assets/78b71cac-4e00-4423-ba53-83fd44ab6cba)

Reply:
I removed all `from qibo.models import Circuit` in `qibo/transpiler`, and it worked. Most of the `Circuit` imports were only used for type hints. In the other cases, I moved `from qibo.models import Circuit` inside the functions.

Reply:
Ok, this is truly a `qibo-core`-related problem once again, since the `Circuit` should be one of the most basic instances, and instead is effectively depending on the backends' machinery through its `.execute()` method...

Thanks for the investigation, I lifted your comment into #1490, you can resolve this conversation.